Returning structs

chriseth opened the pull request:
- [x] change json-ABI
 - [x] change function signature for function identifiers

Fixes #1601.

chriseth committed:
Allow structs as part of function interfaces.

chriseth committed:
Check for recursive structs.

chriseth committed:
Update tests and error messages.

VoR0220 commented on test/libsolidity/SolidityABIJSON.cpp:
```diff
@@ -752,6 +752,140 @@ BOOST_AUTO_TEST_CASE(function_type)
+		"constant" : false,
+		"inputs" : [{
+			"name" : "s",
+			"type" : "L.S storage"
```
Why is storage being denoted here? This would seem to be breaking the ABI type system even by the EIP's standards. I guess my question here is...should the ABI care whether or not data is in storage or in memory, or should that be a user discretion?

chriseth committed:
Function signatures containing structs.

chriseth committed:
Tests for external signatures.

chriseth committed:
Fix interface type conversion internal to structs.

chriseth committed:
Structure type json using "components".

chriseth committed:
Document structs in ABI

chriseth committed:
Expect test to fail until implemented.

VoR0220 commented on docs/abi-spec.rst:
```diff
@@ -343,3 +346,86 @@ would result in the JSON:
+        {
+          "name": "s",
+          "type": "",
+          "components": [
```
Much better! This is a far better scheme...perhaps we should make the type of the type of the contract? I'm unsure about leaving it blank. "Test.S" Maybe would be the struct type? Either way, this is a no doubt improvement.

VoR0220 commented on docs/abi-spec.rst:
```diff
@@ -343,3 +346,86 @@ would result in the JSON:
+            },
+            {
+              "name": "c",
+              "type": "[]",
```
Is it possible to name the struct type here as well? 

A potentially better scheme overall might be to define the structs at the contract level so that one knows what they're interacting with when parsing for components of X? 

For example, it could be possible to define Contract Test in this instance (abi encoded wise) as:

```json
{
"Test": [{
   "type": "struct",
   "name": "T",
   "components": [
          {
               "name": "x",
               "type": "uint256"
          },
          {
                 "name": "y",
                  "type":"uint256"
          },
    ]
},
{
  "type": "struct",
  "name": "S",
  "components": [
        {
            "name": "a", 
            "type": "uint256" 
        },
        {
             "name": "b",
             "type": "uint[]"
        },
       {
             "name": "c",
             "type": "T[]"
       }
    ]
},
{
      "name": "f", 
      "type": "function",
      "inputs": [
             {
                  "type": "Test.S",
                  "name": "s"
             }
             {
                  "name": "t",
                  "type": "Test.T"
             }
             {
                  "name": "a",
                  "type": "uint256"
             }
       ]
}
]
```

The pros of this approach is that it's cleaner and better able to be parsed into a type/object of type when it reaches "client side" code for interaction (js, go, rust, whatever you're developing with). It also makes this an easier setup in terms of a recursive read of embedded structs within the struct type. The cons of this come whenever it comes to external referencing to another contract (you would likely need the abi of the contract holding information about the external/embedded type).

One solution to this might be to define structs into two groups: "internal structs", and "external structs", meaning it will copy a reference of the contract and type in there along with the hash of the metadata of the contract it references. Solidity could then safely ensure that the struct belongs to the external contract before passing data outwards.

chriseth replied:
The ABI does not have a concept of types being named. Because of that, I would not introduce the name of the struct behind the tuple type. The idea behind naming the elements of the tuple type is that the return value can be transformed into a json value with named components.

axic commented on libsolidity/ast/Types.h:
```diff
@@ -241,9 +241,15 @@ class Type: private boost::noncopyable, public std::enable_shared_from_this<Type
+	/// @returns the canonical name of this type for use in library function signatures.
+	virtual std::string canonicalName() const { return toString(true); }
+	/// @returns the signature of this type in external functions, i.e. `uint256` for integers
+	/// or `(uint256, bytes8)[2]` for an array of structs. If @a _structsByName,
```
Is there a space between the types in the sig?

axic commented on test/libsolidity/SolidityABIJSON.cpp:
```diff
@@ -46,7 +46,7 @@ class JSONInterfaceChecker
 		Json::Value generatedInterface = m_compilerStack.contractABI("");
 		Json::Value expectedInterface;
-		m_reader.parse(_expectedInterfaceString, expectedInterface);
+		BOOST_REQUIRE(m_reader.parse(_expectedInterfaceString, expectedInterface));
```
I think this would throw if an error occurs so it should use `ETH_TEST_REQUIRE_NO_THROW`.

chriseth replied:
If a test throws an exception, it fails automatically. The `ETH_TEST_REQUIRE_NO_THROW` was only introduced to improve the error message, but it has some other nasty drawbacks.

axic commented:
Rebased.

axic commented on libsolidity/ast/Types.h:
```diff
@@ -757,6 +757,10 @@ class StructType: public ReferenceType
 	std::set<std::string> membersMissingInMemory() const;
 
+	/// @returns true if the same struct is used recursively in one of its members. Only
+	/// analyziz the "memory" representation, i.e. mappings are ignored in all structs.
```
"analyziz"

chriseth committed:
Fix tests.

chriseth committed:
Implement struct encoder.

chriseth commented:
This should be ready now.

chriseth commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -30,13 +30,6 @@ using namespace std;
-{
-	// This throws an exception and thus might cause immediate termination, but hey,
-	// it's a failed assertion anyway :-)
-	solAssert(m_requestedFunctions.empty(), "Forgot to call ``requestedFunctions()``.");
```
This is not really useful, it will break hard if this assert is hit. Also, it prevents regular exceptions in this module from being propagated, and finally, code that has missing functions will not compile anyway.

axic commented:
Does this now obsoletes #2433?

chriseth commented:
Yes, but the text in the description of #2433 should probably be turned into new issues.

axic commented on docs/abi-spec.rst:
```diff
@@ -345,3 +348,86 @@ would result in the JSON:
+      "inputs": [
+        {
+          "name": "s",
+          "type": "",
```
Shouldn't this state `"struct"`?

axic replied:
Or `tuple`.

chriseth replied:
ok, so I will change this to `tuple` everywhere, but it can also be `tuple[][4]` or similar.

chriseth replied:
ok, actually ``struct`` is better than ``tuple`` because components in tuples are usually not named.

axic commented on docs/abi-spec.rst:
```diff
@@ -310,6 +312,7 @@ An event description is a JSON object with fairly similar fields:
 - `inputs`: an array of objects, each of which contains:
   * `name`: the name of the parameter;
   * `type`: the canonical type of the parameter.
+  * `components`: used for tuple types (more below).
```
If you with `struct` then I'd add a comment here to mean `tuple` or `struct` types.

axic commented on docs/abi-spec.rst:
```diff
@@ -345,3 +348,86 @@ would result in the JSON:
   "outputs": []
   }]
+
+Use of Structs in Types
```
And perhaps make a generic section about "tuple" encoding, where `name` is optional in case of tuples and is mandatory in case of structs (aka named tuples).

chriseth replied:
The thing is that the type itself does not have named components, the names only appear in the JSON description of the interface. I can remove the term "tuple" from the documentation if that makes it easier to understand.

chriseth committed:
Use "tuple" for struct types in ABI JSON.

Only use tuple as a type in the ABI (and remove all "anonymous struct" references too)

chriseth commented on test/libsolidity/SolidityNameAndTypeResolution.cpp:
```diff
@@ -599,6 +599,61 @@ BOOST_AUTO_TEST_CASE(enum_external_type)
+			BOOST_CHECK_EQUAL("f(Test.ActionChoices,uint256,Test.Empty)", functions[0]->externalSignature());
+			BOOST_CHECK_EQUAL("g(Test,Test.Nested)", functions[1]->externalSignature());
+			BOOST_CHECK_EQUAL("h(function[])", functions[2]->externalSignature());
+			BOOST_CHECK_EQUAL("i(Test.Nested[])", functions[3]->externalSignature());
```
Since passing storage pointers to structs has been possible for a while with libraries, but it used this notation to compute the function signature, we now have to decide whether we want to change it for memory structs passed to libraries. I would say we just stay with this notation, since it is more details in other ways, too (for example, enums and contracts are passed by name, too).

VoR0220 replied:
what are the pros and cons of each approach?

axic commented:
Rebased. I think we need to finish the documentation part and should fein to merge otherwise.

axic committed:
Clarify ABI & Solidity types

axic committed:
Add changelog

axic commented:
Rebased.

@chriseth please check the changelog.

The middle commit of the tuple->struct->tuple change could be removed too :)

axic commented:
First case fails as expected, though the message could be improved:
```
pragma experimental ABIEncoderV2;

contract C {
  struct S1 { }
  struct S2 { }
  function f(S1) {}
  function f(S2) {}
}
```
->
```
struct.sol:7:3: Error: Function overload clash during conversion to external types for arguments.
  function f(S2) {}
  ^---------------^
```

```
pragma experimental ABIEncoderV2;

contract C {
  struct S1 { function () a; }
  struct S2 { bytes24 a; }
  function f(S1) {}
  function f(S2) {}
}
```
->
```
Internal compiler error during compilation:
/Users/alex/Projects/solidity/libsolidity/ast/Types.cpp(1819): Throw in function auto dev::solidity::StructType::signatureInExternalFunction(bool)::(anonymous class)::operator()(TypePointer) const
Dynamic exception type: boost::exception_detail::clone_impl<dev::solidity::InternalCompilerError>
std::exception::what: 
[dev::tag_comment*] = 
```

chriseth committed:
Check for interface types of members and cache recursion check.

axic commented:
Please add all these and the below as test cases:
```
pragma experimental ABIEncoderV2;

contract C {
  struct S1 { function () external a; }
  struct S2 { bytes24 a; }
  function f(S1) {}
  function f(S2) {}
}
```

```
contract C {
  struct S { mapping(string => string) a; }
  function f(S) {}
}
```

chriseth committed:
Fixed tests with mappings in structs and added some more.

chriseth commented:
Wohoooo!
